pkg/plugin/processor/standalone/host_module.go

```diff
@@ -139,9 +155,137 @@ func (m *hostModuleInstance) commandResponse(ctx context.Context, buf types.Byte
+		return pprocutils.ErrorCodeUnknownCommandResponse
 	}
 
 	m.commandResponses <- tuple[*processorv1.CommandResponse, error]{&resp, nil}
 	return 0
 }
+
+// createSchema is the exported function that is called by the WASM module to
+// create a schema and set the buffer bytes to the marshalled response.
+// It returns the response size on success, or an error code on error.
+func (m *hostModuleInstance) createSchema(ctx context.Context, buf types.Bytes) types.Uint32 {
+	m.logger.Trace(ctx).Msg("executing create_schema")
+
+	// check if there's a parked response from the last call
+	if m.parkedCreateSchemaResponse != nil {
+		// make sure the request is the same as the last one
+		if bytes.Equal(buf[:len(m.parkedCreateSchemaBuffer)], m.parkedCreateSchemaBuffer) {
+			out, err := proto.MarshalOptions{}.MarshalAppend(buf[:0], m.parkedCreateSchemaResponse)
+			if err != nil {
+				m.logger.Err(ctx, err).Msg("failed marshalling protobuf create schema response")
+				return pprocutils.ErrorCodeStart
+			}
+			m.parkedCreateSchemaResponse = nil
+			m.parkedCreateSchemaBuffer = nil
+			return types.Uint32(len(out))
+		}
+	}
+
+	var req procutilsv1.CreateSchemaRequest
+	length := len(buf)
+	if m.parkedCreateSchemaBuffer != nil {
+		length = len(m.parkedCreateSchemaBuffer)
+	}
+	err := proto.Unmarshal(buf[:length], &req)
+	if err != nil {
+		m.logger.Err(ctx, err).Msg("failed unmarshalling protobuf create schema request")
+		return pprocutils.ErrorCodeStart
+	}
+
+	var pErr *pprocutils.Error
+	schemaResp, err := m.schemaService.CreateSchema(ctx, fromproto.CreateSchemaRequest(&req))
+	if err != nil {
+		if cerrors.As(err, &pErr) {
+			return types.Uint32(pErr.ErrCode)
+		}
+		return pprocutils.ErrorCodeStart
+	}
+	resp := toproto.CreateSchemaResponse(schemaResp)
+	m.parkedCreateSchemaResponse = resp
+	m.parkedCreateSchemaBuffer = buf
+
+	// If the buffer is too small, we park the response and return the size of the
+	// response. The next call to createSchema will return the same response.
+	if size := proto.Size(m.parkedCreateSchemaResponse); len(buf) < size {
+		m.logger.Warn(ctx).
+			Int("response_bytes", size).
+			Int("allocated_bytes", len(buf)).
+			Msgf("insufficient memory, response will be parked until next call to create_schema")
+		return types.Uint32(size)
+	}
+	out, err := proto.MarshalOptions{}.MarshalAppend(buf[:0], m.parkedCreateSchemaResponse)
+	if err != nil {
+		m.logger.Err(ctx, err).Msg("failed marshalling protobuf create schema response")
+		return pprocutils.ErrorCodeStart
+	}
+	m.parkedCreateSchemaResponse = nil
+	m.parkedCreateSchemaBuffer = nil
+
+	return types.Uint32(len(out))
+}
+
+// getSchema is the exported function that is called by the WASM module to
+// get a schema and set the buffer bytes to the marshalled response.
+// It returns the response size on success, or an error code on error.
+func (m *hostModuleInstance) getSchema(ctx context.Context, buf types.Bytes) types.Uint32 {
```

Comment:
getSchema and createSchema work almost identically (check for parked response, get request, unmarshall, do something with it, etc.). WDYT about trying to refactoring this into "something" (a struct or a function) that does that?

Reply:
Makes sense, similarly to how we extracted the utility function in the SDK 👍

Reply:
thought about that, but the problem is with the number of different variables needed and functions and types, so I don't feel like it would look pretty, but will give it a go